src/picongpu/include/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp

```diff
@@ -264,29 +267,27 @@ namespace ionization
                 ValueType_Ene kinEnergyV = Field2ParticleInterpolation()
                     (cachedEne.shift(localCell).toCursor(), pos, fieldPosEne());
 
-                /* define number of bound macro electrons before ionization */
-                float_X prevBoundElectrons = particle[boundElectrons_];
-
                 /* density in sim units */
-                float_X density = densityV[0];
+                float_X const density = densityV[0];
                 /* energy density in sim units */
-                float_X kinEnergyDensity = kinEnergyV[0];
+                float_X const kinEnergyDensity = kinEnergyV[0];
 
-                /* this is the point where actual ionization takes place */
+                /* Returns the new number of bound electrons for an integer number of macro electrons */
                 IonizationAlgorithm ionizeAlgo;
-                float_X newBoundElectrons = ionizeAlgo(
-                    kinEnergyDensity, density,
-                    particle, this->randomGen()
-                    );
-
-                particle[boundElectrons_] = newBoundElectrons;
-
-                /* safety check to avoid double counting since recombination is not yet implemented */
-                if (prevBoundElectrons > newBoundElectrons)
-                {
-                    /* determine number of new macro electrons to be created */
-                    newMacroElectrons = static_cast<unsigned int>(prevBoundElectrons - newBoundElectrons);
-                }
+                numNewFreeMacroElectrons = ionizeAlgo(
+                    kinEnergyDensity,
+                    density,
+                    particle,
+                    this->randomGen()
+                );
+
+                /* ionization of the ion by reducing the number of bound electrons
+                 *
+                 * optimization: only accesses global memory if the charge
+                 *               state did really change
+                 */
+                if( numNewFreeMacroElectrons > 0u )
+                    particle[ boundElectrons_ ] -= float_X( numNewFreeMacroElectrons );
```

Comment:
**offline:** Since we always check beforehand if `chargeState < protonNumber` we prevent that from happening and we can leave the redesign open until we do it for all the ionization models.
We could put a `\todo` here until it is fixed by design with issue #1850 
```c++
/*
 * \todo refactor for consistency with the other ionization methods
 *    and to preserve a valid ion state (issue #1850)
 */